## 팀원
- 김장현 / 컴퓨터공학부/ 2019-26471 
- 안형서 / 전기정보공학부 / 2017-12932
- 양서연 / 인공지능 협동과정 / 2021-33205
- 이욱재 / 산업공학과 / 2021-22053

## Intro 
- Here, we train neural networks solving the four Korean language tasks ([link](https://corpus.korean.go.kr/task/taskDownload.do?taskId=1&clCd=END_TASK&subMenuId=sub02)). 
- Our basic approach is to fine-tune pre-trained **korean language models** (https://huggingface.co/models?language=ko&sort=downloads). 
- We basically use **KLUE-RoBERTa** models from https://github.com/KLUE-benchmark/KLUE, which is the state-of-the-art korean language model. 
- We refer the following sources for the some parts of data processing and fine-tuning techniques. 
 - Sun et al., 'How to Fine-Tune BERT for Text Classification?', https://arxiv.org/abs/1905.05583
 - https://github.com/NIKL-Team-BC/NIKL-KLUE
- For more **detailed codes and experiment logs**, please refer to our [github page](https://github.com/codestella/nlp_final_project)

### Disclaimer
코랩, 리눅스 환경 쥬피터 등 다양한 환경의 쥬피터 파일들을 하나로 합치는 과정에서, 호환이 안 맞아 생기는 문제가 있습니다:   
```Notebook validation failed: Additional properties are not allowed ('metadata' was unexpected):```

코드가 돌아가는데에 문제는 없음을 확인했으니, 일단 무시하고 진행해서 코드 체크를 하시면 될 것 같습니다. 
혹여 문제가 있다면 개별 쥬피터 파일을 모아놓은 https://github.com/codestella/nlp_final_project/tree/main/jupyter 를 참조 바랍니다.

학습된 모델은 요청시에 제공 가능합니다.

## 1. 판정의문문 (BoolQ)

아래 하이퍼파라미터 및 모델 세팅을 탐색 하였음. 가장 성능이 좋은 세팅은 강조되어 있음. 이 실험에 대한 보다 자세한 코드 및 로그파일은 [github page](https://github.com/codestella/nlp_final_project) 참조. 
- Model size: **Large** (\~85%) / Base (\~79%)
- Epoch: 10
- warm up: 10% training step (없으면 학습 불안정)
- Learning rate: 1e-5, **8e-6**, 5e-6 (큰 차이는 없으나, 커지면 학습 불안정)
- Batch size: **5**, 20, 60
- Finetuning: **All**, Only classifier (i.e., freeze feature extractor, \~57%)
- Classifier: linear model 충분 (multi-layer로 늘려도 gain 작음)

In [1]:
import os
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
import transformers
import pandas
from transformers import AutoTokenizer
from transformers import RobertaModel, RobertaConfig
from transformers import AdamW
import time
import argparse

transformers.logging.set_verbosity(40) # Turn off warning
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

In [2]:
def load_data(path, tokenizer):
    ''' Tokenization for BoolQ data'''
    dataset = pandas.read_csv(path,
                              delimiter='\t',
                              names=['ID', 'text', 'question', 'answer'],
                              header=0)

    tokenized = tokenizer(dataset['text'].tolist(),
                          dataset['question'].tolist(),
                          padding=True,
                          truncation=True,
                          return_tensors="pt")
    dataset['label'] = torch.tensor(dataset['answer'])
    return dataset, tokenized


class Roberta(RobertaModel):
    ''' Classification layer added Roberta model'''
    def __init__(self, config, model_name):
        super(Roberta, self).__init__(config)
        self.roberta = RobertaModel.from_pretrained(model_name, config=config)
        self.hdim = config.hidden_size
        self.nclass = config.nclass
        self.classifier = nn.Linear(self.hdim, self.nclass)

    def forward(self, input_ids, attention_mask, **kwargs):
        outputs = self.roberta(input_ids, attention_mask=attention_mask)
        h = outputs[0][:, 0, :]
        logits = self.classifier(h)
        return logits

In [3]:
class TensorDataset(Dataset):
    ''' Define Torch Dataset '''
    def __init__(self, tokenized_dataset, labels):
        self.tokenized_dataset = tokenized_dataset
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: val[idx] for key, val in self.tokenized_dataset.items()}
        label = self.labels[idx]
        return item, label

    def __len__(self):
        return len(self.labels)

In [4]:
# Define tokenizer and model type
model_type = "Roberta"
size = 'large'
model_name = f"klue/roberta-{size}"
tokenizer = AutoTokenizer.from_pretrained(model_name)

Set data path below!

In [5]:
# Data path
base_path = './data'

train_dataset, train_tokenized = load_data(os.path.join(base_path, 'SKT_BoolQ_Train.tsv'), tokenizer)
val_dataset, val_tokenized = load_data(os.path.join(base_path, 'SKT_BoolQ_Dev.tsv'), tokenizer)

train_dataset = TensorDataset(train_tokenized, train_dataset['label'])
val_dataset = TensorDataset(val_tokenized, val_dataset['label'])

# Define loader
if size == 'base':
    batch_size = 16
else:
    batch_size = 5
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)

In [6]:
# Data example
tokenizer.decode(train_tokenized['input_ids'][0])

'[CLS] 로마 시대의 오리엔트의 범위는 제국 내에 동부 지방은 물론 제국 외부에 있는 다른 국가에 광범위하게 쓰이는 단어였다. 그 후에 로마 제국이 분열되고 서유럽이 그들의 중심적인 세계를 형성하는 과정에서 자신들을 옥시덴트 ( occident ), 서방이라 부르며 오리엔트는 이와 대조되는 문화를 가진 동방세계라는 뜻이 부가되어, 인도와 중국, 일본을 이루는 광범위한 지역을 지칭하는 단어가 되었다. [SEP] 오리엔트는 인도와 중국, 일본을 이루는 광범위한 지역을 지칭하는 단어로 쓰인다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [P

### [BoolQ] 학습 결과
- 최적 세팅에서 10번 반복한 결과 84% \~ 87% 의 결과를 얻음
- 각 실험은 1시간 정도 소요 (6min/epoch) 
- (참고) jupyter를 서버에서 돌려서 컴퓨터 연결이 끊겼을때 print가 잘 되지 않은 경우가 있으나, 모델 학습 및 저장은 이상 없음.
- 본 모델을 앙상블하여 최종 **88.14\%**의 validation accuracy 달성.

In [7]:
def train_epoch(epoch, model, train_loader, optimizer, scheduler):
    ''' One epoch fine-tuning '''
    model.train()
    total_loss = 0
    cor = 0
    n_sample = 0
    s = time.time()
    criterion = nn.CrossEntropyLoss()

    for data, target in train_loader:
        item = {key: val.to(device) for key, val in data.items()}
        target = target.to(device)

        logits = model(**item)
        loss = criterion(logits, target)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        with torch.no_grad():
            preds = torch.argmax(logits, dim=-1)

        total_loss += loss.item()
        cor += (preds == target).sum().item()
        n_sample += len(target)

        print(f"{cor}/{n_sample}", end='\r')

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    print(
        f"[Epoch {epoch}] Train loss: {loss_avg:.3f}, acc: {acc*100:.2f}, time: {time.time()-s:.1f}s"
    )
    return acc


def validate(epoch, model, val_loader, verbose=True):
    ''' Evaluate on validation set '''
    model.eval()
    total_loss = 0
    cor = 0
    n_sample = 0
    criterion = nn.CrossEntropyLoss()
    pred_all = []
    
    with torch.no_grad():
        for data, target in val_loader:
            item = {key: val.to(device) for key, val in data.items()}
            target = target.to(device)

            logits = model(**item)
            loss = criterion(logits, target)
            preds = torch.argmax(logits, dim=-1)
            pred_all.append(preds)

            total_loss += loss.item()
            cor += (preds == target).sum().item()
            n_sample += len(target)

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    pred_all = torch.cat(pred_all)
    
    if verbose:
        print(f"[Epoch {epoch}] Valid loss: {loss_avg:.3f}, acc: {acc*100:.2f}")
    return acc, pred_all


def train(idx, num_epochs, lr, train_loader, val_loader, config, save_dir='./results'):
    ''' Train for multiple epochs and validate '''    
    print(f"Start trining {idx}th model")
    model = Roberta(config, model_name).to(device)
    optimizer = AdamW(model.parameters(), lr=lr)
    scheduler = transformers.get_scheduler("linear",
                                           optimizer=optimizer,
                                           num_warmup_steps=num_epochs * len(train_loader) // 10,
                                           num_training_steps=num_epochs * len(train_loader))
    best_acc = 0
    for epoch in range(num_epochs):
        train_acc = train_epoch(epoch, model, train_loader, optimizer, scheduler)
        val_acc, _ = validate(epoch, model, val_loader)
        if val_acc > best_acc:
            best_acc = val_acc

            model_to_save = model.module if hasattr(model, "module") else model
            model_to_save.save_pretrained(os.path.join(save_dir, f'{idx}'))
            
    print(f"Training finish! Best validation accuracy: {best_acc*100:.2f}\n")

In [8]:
def validate_ensemble(save_dir, model_name, val_loader, answer, idx_max=10, acc_threshold=0.85):
    ''' Measure ensemble accuracy '''
    pred_ensemble = []
    for idx in range(idx_max):
        model = Roberta.from_pretrained(os.path.join(save_dir, f'{idx}'), model_name)
        model.to(device)
        acc, pred_all = validate('best', model, val_loader, verbose=False)
        print(f"Load {idx}th model (acc: {acc*100:.2f})")
        if acc >= acc_threshold:
            pred_ensemble.append(pred_all)
        
    pred_ensemble = torch.stack(pred_ensemble, dim=-1).float()
    pred_ensemble = (pred_ensemble.mean(-1) >= 0.5).long().to(answer.device)
    acc_ensemble = (pred_ensemble == answer).sum() / len(answer)
    print(f"\nEnsemble accuracy: {acc_ensemble*100:.2f}")

In [9]:
# Fine-tuning networks (10 repeat)
lr = 8e-6
num_epochs = 10

save_dir = './results_qa'
if not os.path.exists(save_dir):
    os.makedirs(save_dir)
    
config = RobertaConfig.from_pretrained(model_name)
config.nclass = 2
for i in range(10):
    train(i, num_epochs, lr, train_loader, val_loader, config=config, save_dir=save_dir)

Start trining 0th model


Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

[Epoch 0] Train loss: 0.135, acc: 55.66, time: 346.7s
[Epoch 0] Valid loss: 0.111, acc: 72.00
[Epoch 1] Train loss: 0.086, acc: 81.53, time: 348.5s
[Epoch 1] Valid loss: 0.090, acc: 78.71
[Epoch 2] Train loss: 0.037, acc: 93.40, time: 350.1s
[Epoch 2] Valid loss: 0.094, acc: 82.57
[Epoch 3] Train loss: 0.016, acc: 97.33, time: 349.0s
[Epoch 3] Valid loss: 0.112, acc: 84.00
[Epoch 4] Train loss: 0.006, acc: 98.99, time: 347.3s
[Epoch 4] Valid loss: 0.167, acc: 84.00
[Epoch 5] Train loss: 0.004, acc: 99.40, time: 350.4s
[Epoch 5] Valid loss: 0.165, acc: 84.00
[Epoch 6] Train loss: 0.003, acc: 99.48, time: 350.7s
[Epoch 6] Valid loss: 0.144, acc: 87.00
[Epoch 7] Train loss: 0.001, acc: 99.81, time: 349.5s
[Epoch 7] Valid loss: 0.181, acc: 85.71
[Epoch 8] Train loss: 0.001, acc: 99.89, time: 353.4s
[Epoch 8] Valid loss: 0.189, acc: 85.00
[Epoch 9] Train loss: 0.001, acc: 99.89, time: 350.9s
[Epoch 9] Valid loss: 0.193, acc: 84.43
Training finish! Best validation accuracy: 87.00

Start trin

Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

[Epoch 0] Train loss: 0.130, acc: 59.15, time: 351.2s
[Epoch 0] Valid loss: 0.101, acc: 75.86
[Epoch 1] Train loss: 0.081, acc: 82.18, time: 349.3s
[Epoch 1] Valid loss: 0.070, acc: 84.71
[Epoch 2] Train loss: 0.031, acc: 94.22, time: 350.0s
[Epoch 2] Valid loss: 0.126, acc: 84.71
[Epoch 3] Train loss: 0.013, acc: 97.93, time: 354.0s
[Epoch 3] Valid loss: 0.132, acc: 84.29
[Epoch 4] Train loss: 0.007, acc: 99.07, time: 353.1s
[Epoch 4] Valid loss: 0.152, acc: 84.57
[Epoch 5] Train loss: 0.002, acc: 99.54, time: 353.0s
[Epoch 5] Valid loss: 0.164, acc: 84.71
[Epoch 6] Train loss: 0.003, acc: 99.54, time: 352.5s
[Epoch 6] Valid loss: 0.150, acc: 86.14
[Epoch 7] Train loss: 0.002, acc: 99.73, time: 351.3s
[Epoch 7] Valid loss: 0.168, acc: 85.71
[Epoch 8] Train loss: 0.001, acc: 99.97, time: 354.4s
[Epoch 8] Valid loss: 0.179, acc: 85.29
[Epoch 9] Train loss: 0.001, acc: 99.95, time: 353.6s
[Epoch 9] Valid loss: 0.179, acc: 85.71
Training finish! Best validation accuracy: 86.14

Start trin

Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

[Epoch 0] Train loss: 0.140, acc: 50.97, time: 353.7s
[Epoch 0] Valid loss: 0.134, acc: 59.14
[Epoch 1] Train loss: 0.104, acc: 73.29, time: 353.2s
[Epoch 1] Valid loss: 0.083, acc: 81.14


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 4] Train loss: 0.006, acc: 99.13, time: 352.3s
[Epoch 4] Valid loss: 0.148, acc: 83.00
[Epoch 5] Train loss: 0.004, acc: 99.29, time: 354.9s
[Epoch 5] Valid loss: 0.143, acc: 83.43
[Epoch 6] Train loss: 0.003, acc: 99.56, time: 353.8s
[Epoch 6] Valid loss: 0.193, acc: 82.86
[Epoch 7] Train loss: 0.000, acc: 99.95, time: 356.7s
[Epoch 7] Valid loss: 0.202, acc: 84.71


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 5] Train loss: 0.002, acc: 99.70, time: 353.3s
[Epoch 5] Valid loss: 0.179, acc: 85.00
[Epoch 6] Train loss: 0.002, acc: 99.59, time: 353.9s
[Epoch 6] Valid loss: 0.188, acc: 84.86
[Epoch 7] Train loss: 0.001, acc: 99.84, time: 355.2s
[Epoch 7] Valid loss: 0.184, acc: 85.00
[Epoch 8] Train loss: 0.001, acc: 99.92, time: 355.2s
[Epoch 8] Valid loss: 0.184, acc: 85.86


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 2] Train loss: 0.035, acc: 94.00, time: 354.7s
[Epoch 2] Valid loss: 0.100, acc: 82.00
[Epoch 3] Train loss: 0.012, acc: 98.25, time: 357.2s
[Epoch 3] Valid loss: 0.128, acc: 83.57
[Epoch 4] Train loss: 0.007, acc: 98.80, time: 354.8s
[Epoch 4] Valid loss: 0.150, acc: 82.57
[Epoch 5] Train loss: 0.005, acc: 98.99, time: 356.3s
[Epoch 5] Valid loss: 0.187, acc: 81.43


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 7] Train loss: 0.002, acc: 99.75, time: 358.5s
[Epoch 7] Valid loss: 0.181, acc: 83.57
[Epoch 8] Train loss: 0.001, acc: 99.92, time: 358.0s
[Epoch 8] Valid loss: 0.193, acc: 82.57
[Epoch 9] Train loss: 0.000, acc: 99.97, time: 356.7s
[Epoch 9] Valid loss: 0.209, acc: 81.86
Training finish! Best validation accuracy: 83.57

Start trining 5th model


Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

[Epoch 0] Train loss: 0.140, acc: 51.24, time: 357.3s
[Epoch 0] Valid loss: 0.123, acc: 68.14


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 9] Train loss: 0.001, acc: 99.95, time: 359.1s
[Epoch 9] Valid loss: 0.218, acc: 84.43
Training finish! Best validation accuracy: 84.86

Start trining 6th model


Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

[Epoch 0] Train loss: 0.137, acc: 53.12, time: 359.3s
[Epoch 0] Valid loss: 0.098, acc: 76.57
[Epoch 1] Train loss: 0.086, acc: 81.17, time: 354.8s
[Epoch 1] Valid loss: 0.086, acc: 81.86
[Epoch 2] Train loss: 0.033, acc: 94.00, time: 358.5s
[Epoch 2] Valid loss: 0.113, acc: 81.14


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 6] Train loss: 0.002, acc: 99.86, time: 356.8s
[Epoch 6] Valid loss: 0.181, acc: 84.57
[Epoch 7] Train loss: 0.001, acc: 99.89, time: 358.8s
[Epoch 7] Valid loss: 0.199, acc: 84.57
[Epoch 8] Train loss: 0.000, acc: 99.95, time: 357.7s
[Epoch 8] Valid loss: 0.205, acc: 85.29
[Epoch 9] Train loss: 0.000, acc: 100.00, time: 355.7s
[Epoch 9] Valid loss: 0.209, acc: 85.43
Training finish! Best validation accuracy: 85.43

Start trining 7th model


Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 1] Train loss: 0.085, acc: 80.90, time: 355.3s
[Epoch 1] Valid loss: 0.097, acc: 81.00
[Epoch 2] Train loss: 0.037, acc: 93.70, time: 356.8s
[Epoch 2] Valid loss: 0.088, acc: 84.57
[Epoch 3] Train loss: 0.013, acc: 98.25, time: 355.4s
[Epoch 3] Valid loss: 0.115, acc: 84.57
[Epoch 4] Train loss: 0.009, acc: 98.64, time: 357.4s
[Epoch 4] Valid loss: 0.121, acc: 85.00


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 3] Train loss: 0.015, acc: 97.44, time: 357.1s
[Epoch 3] Valid loss: 0.130, acc: 83.71
[Epoch 4] Train loss: 0.006, acc: 99.02, time: 356.7s
[Epoch 4] Valid loss: 0.135, acc: 84.29
[Epoch 5] Train loss: 0.003, acc: 99.48, time: 356.5s
[Epoch 5] Valid loss: 0.177, acc: 84.00
[Epoch 6] Train loss: 0.001, acc: 99.95, time: 359.4s
[Epoch 6] Valid loss: 0.232, acc: 84.00


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 9] Train loss: 0.001, acc: 99.86, time: 361.4s
[Epoch 9] Valid loss: 0.224, acc: 83.71
Training finish! Best validation accuracy: 84.29

Start trining 9th model


Some weights of the model checkpoint at klue/roberta-large were not used when initializing RobertaModel: ['lm_head.decoder.weight', 'lm_head.layer_norm.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.decoder.bias', 'lm_head.layer_norm.weight', 'lm_head.bias']
- This IS expected if you are initializing RobertaModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaModel were not initialized from the model checkpoint at klue/roberta-large and are newly initialized: ['roberta.pooler.dense.weight', 'roberta.pooler.dense.bias']
You should probably TRAIN this model on a down-stream task to be able to use it f

[Epoch 0] Train loss: 0.141, acc: 51.65, time: 359.3s
[Epoch 0] Valid loss: 0.153, acc: 46.57
[Epoch 1] Train loss: 0.104, acc: 74.05, time: 357.5s
[Epoch 1] Valid loss: 0.088, acc: 79.43
[Epoch 2] Train loss: 0.048, acc: 90.86, time: 356.8s
[Epoch 2] Valid loss: 0.084, acc: 82.71
[Epoch 3] Train loss: 0.017, acc: 97.57, time: 357.0s
[Epoch 3] Valid loss: 0.150, acc: 79.57


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



[Epoch 4] Train loss: 0.007, acc: 98.83, time: 360.3s
[Epoch 4] Valid loss: 0.162, acc: 83.14
[Epoch 5] Train loss: 0.004, acc: 99.32, time: 360.5s
[Epoch 5] Valid loss: 0.185, acc: 83.57
[Epoch 6] Train loss: 0.004, acc: 99.51, time: 360.3s
[Epoch 6] Valid loss: 0.161, acc: 84.29
[Epoch 7] Train loss: 0.003, acc: 99.62, time: 360.3s
[Epoch 7] Valid loss: 0.164, acc: 83.86


IOPub message rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_msg_rate_limit`.

Current values:
NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [10]:
answer = torch.tensor(val_dataset.labels)
save_dir = './results_qa'

validate_ensemble(save_dir, model_name, val_loader, answer, idx_max=10, acc_threshold=0.85)

Load 0th model (acc: 87.00)
Load 1th model (acc: 86.14)
Load 2th model (acc: 85.43)
Load 3th model (acc: 85.86)
Load 4th model (acc: 83.57)
Load 5th model (acc: 84.86)
Load 6th model (acc: 85.43)
Load 7th model (acc: 85.43)
Load 8th model (acc: 84.29)
Load 9th model (acc: 84.29)

Ensemble accuracy: 88.14


## 2. 동형이의어 구별 (WIC)

- 전체적인 코드의 뼈대 / 하이퍼 파라미터는 같은 조의 BoolQ의 코드를 사용하였습니다.
- 사용모델 : KLUE-roberta-large
- Epoch: 10
- warm up: 10% training step
- Learning rate: 8e-6
- Batch size: 4 (colab에서 4를 넘어가면 memory 초과)
- Classifier: 1 layer linear model ( layer/activaation을 추가해도 효과 미미)

In [2]:
import os
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
import transformers
import pandas
from transformers import AutoTokenizer
from transformers import RobertaModel, RobertaConfig ,RobertaPreTrainedModel
from transformers import AdamW
import time
import argparse
from tqdm.auto import tqdm
import numpy as np
import random

transformers.logging.set_verbosity(40) # Turn off warning

In [4]:
save_dir = './result_wic'

if not os.path.exists( save_dir):
    os.makedirs(save_dir)

### [WIC] 전처리 / 데이터셋 만들기
 - 동형이의어의 토큰이 다른 단어와 섞이지 않고 명확하게 쪼개져야 하기때문에 앞뒤로 special token을 삽입해 줌으로써 구분해줍니다.
 - 또한, 모델의 끝부분에서 동형이의어의 embedding vector만 사용하고, 나머지 토큰들의 결과는 사용하지 않으니, 그 부분만 1, 나머지는 0 인 wordmask가 필요합니다. 이를 위해 만든 speical token의 위치를 파악하여 wordmask를 만들어줍니다.
 - 처음 시도는 torch.roll(shift 연산)을 사용해 동형이의어의 끝 부분을 한칸 앞으로 당긴곳을 1로 만들었지만, 이 경우, 동형이의어가 2개이상의 토큰으로 나뉜경우 완전히 커버할 수 없습니다.
 - 따라서 torch.cumsum(누적합 연산)을 사용하여 동형이의어의 시작/끝부분 사이를 토큰을 전부 1이되도록 만들어주었습니다.
 - 이후 모델 안에서는 wordmask가 1인 위치의 embedding vector을 전부 더해서 classifier에 넣어줍니다.
 - 각 문장당 동형이의어를 대표하는 하나의 vector가 필요하기 때문에, 문장당 wordmask를 만들어주어 총 2개의 wordmask를 만들어주었습니다.

In [6]:
def load_data(path, tokenizer):
    dataset = pandas.read_csv(path,
                              delimiter='\t',
                              names=['ID', 'Target', 'text1', 'text2','answer','s_s1','e_s1','s_s2','e_s2'],
                              header=0)

    dataset_text1=[]
    dataset_text2=[]

    for i,text1 in enumerate(dataset['text1']):   #동형이의어 앞뒤로 special token 추가
        text= text1[:dataset["s_s1"][i]]+"[WORD1S]"+text1[dataset["s_s1"][i]:dataset["e_s1"][i]]+"[WORD1E]"+text1[dataset["e_s1"][i]:]
        dataset_text1.append(text)

    for i,text2 in enumerate(dataset['text2']):
        text= text2[:dataset["s_s2"][i]]+"[WORD2S]"+text2[dataset["s_s2"][i]:dataset["e_s2"][i]]+"[WORD2E]"+text2[dataset["e_s2"][i]:]
        dataset_text2.append(text)
    
    word1s_tok_idx=tokenizer.encode("[WORD1S]")[1]
    word1e_tok_idx=tokenizer.encode("[WORD1E]")[1]
    word2s_tok_idx=tokenizer.encode("[WORD2S]")[1]
    word2e_tok_idx=tokenizer.encode("[WORD2E]")[1]

    tokenized = tokenizer(dataset_text1,
                          dataset_text2,
                          padding=True,
                          truncation=True,
                          return_tensors="pt")
    dataset['label'] = torch.tensor(dataset['answer'],dtype=int)

    #word_mask1=  torch.roll((tokenized["input_ids"]==word1e_tok_idx),-1,dims=1).int()  #first trial with roll
    #word_mask2=  torch.roll((tokenized["input_ids"]==word2e_tok_idx),-1,dims=1).int()
    
    word_mask1 = torch.roll((tokenized["input_ids"]==word1s_tok_idx),1,dims=1).int() # second trial with cumsum
    word_mask1 = word_mask1 - (tokenized["input_ids"]==word1e_tok_idx).int()
    word_mask1 = torch.cumsum(word_mask1,dim=1)

    word_mask2 = torch.roll((tokenized["input_ids"]==word2s_tok_idx),1,dims=1).int()  # 각 문장당 동형이의어의 토큰 위치가 1이되는 mask 생성
    word_mask2 = word_mask2 - (tokenized["input_ids"]==word2e_tok_idx).int()
    word_mask2 = torch.cumsum(word_mask2,dim=1)
    
    return dataset, tokenized, word_mask1, word_mask2


class TensorDataset(Dataset):
    def __init__(self, tokenized_dataset, word_mask1, word_mask2, labels):
        self.tokenized_dataset = tokenized_dataset
        self.word_mask1= word_mask1
        self.word_mask2= word_mask2
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: val[idx] for key, val in self.tokenized_dataset.items()}
        item["word_mask1"]=self.word_mask1[idx]
        item["word_mask2"]=self.word_mask2[idx]
        label = self.labels[idx]
        return item, label

    def __len__(self):
        return len(self.labels)

In [7]:
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

model_type = "Roberta"
size = 'large'
model_name = f"klue/roberta-{size}"
tokenizer = AutoTokenizer.from_pretrained(model_name)

new_special_tokens=["[WORD1S]","[WORD1E]","[WORD2S]","[WORD2E]"]
tokenizer.add_special_tokens({"additional_special_tokens":new_special_tokens})
print(f"added {len(new_special_tokens)} new special token")

print(len(tokenizer))
# add special token - 동형이의어를 문장과 구별하기 위해서

Downloading:   0%|          | 0.00/375 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/243k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/734k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/173 [00:00<?, ?B/s]

added 4 new special token
32004


In [8]:
base_path = './data'

train_dataset, train_tokenized, train_word_mask1, train_word_mask2 = load_data(os.path.join(base_path, 'NIKL_SKT_WiC_Train.tsv'),
                                           tokenizer)
val_dataset, val_tokenized, val_word_mask1, val_word_mask2 = load_data(os.path.join(base_path, 'NIKL_SKT_WiC_Dev.tsv'), tokenizer)

train_dataset = TensorDataset(train_tokenized, train_word_mask1, train_word_mask2, train_dataset['label'])
val_dataset = TensorDataset(val_tokenized, val_word_mask1, val_word_mask2, val_dataset['label'])

if size == 'base':
    batch_size = 16
else:
    batch_size = 4
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)

In [9]:
# 데이터 예시
tokenizer.decode(train_tokenized['input_ids'][0])

'[CLS] 그의 죽음은 타살로 [WORD1S] 단정 [WORD1E] 이 되었다. [SEP] [WORD2S] 단정 [WORD2E] 이 된 교실은 정돈되어 있다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD

In [10]:
# 데이터 예시2, 동형이의어 마스크
train_word_mask1[0]

tensor([0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0])

### [WIC]  Method
- 모델은 KLUE-RoBERTa를 사용하였습니다. (https://github.com/KLUE-benchmark/KLUE) 
- 각 문장의 동형이의어 마스크를 통과한 결과를 합해줌으로써 각 문장의 동형이의어를 대표하는 hid_dim크기의 벡터를 2개 만듭니다.
- 2개의 벡터를 concat하여 classifier에 넣고 구분하였습니다.

**실험 결과**
- 1. **1 layer classifier** (92.54%)
- 2. 2 layer classifier (92.11%)
- 3. 2 layer classifier with relu (91.25%)
- 4. 2 layer classifier with tanh (91.42%)

최종적으로 layer의 복잡성을 올려도, 효과는 미미한데 학습 속도는 느려졌기 때문에 1 layer를 사용하였습니다.

In [11]:
class Roberta(RobertaPreTrainedModel):
    def __init__(self, config, model_name):
        super(Roberta, self).__init__(config)
        self.roberta = RobertaModel.from_pretrained(model_name, config=RobertaConfig.from_pretrained(model_name))
        self.roberta.resize_token_embeddings(config.new_tok_size)
        self.hdim = config.hidden_size
        self.nclass = config.nclass
        self.classifier = nn.Linear(self.hdim*2, self.nclass)
        
        #self.classifier = nn.Linear(self.hdim*2, self.hdim)
        #self.activation = nn.ReLU()
        #self.activation = nn.Tanh()
        #self.classifier2 = nn.Linear(self.hdim, self.nclass) #trial

    def forward(self, input_ids, attention_mask, word_mask1, word_mask2 , **kwargs):
        outputs = self.roberta(input_ids, attention_mask=attention_mask)[0]
        #(batch_size, sequence_length, hidden_size)

        #torch.sum( word_mask1.unsqueeze(2) * outputs , dim=1 ) : word mask가 1인 부분의 embedding vector만 더해서 하나의 벡터로 만듭니다.
        
        h = torch.cat( [torch.sum( word_mask1.unsqueeze(2) * outputs , dim=1 ),torch.sum( word_mask2.unsqueeze(2) * outputs , dim=1 )], dim=1)
        # (batch, hidden_size*2)

        logits = self.classifier(h)

        return logits


config = RobertaConfig.from_pretrained(model_name)
config.nclass = 2
config.new_tok_size = len(tokenizer)
print(config)

Downloading:   0%|          | 0.00/547 [00:00<?, ?B/s]

RobertaConfig {
  "architectures": [
    "RobertaForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 0,
  "classifier_dropout": null,
  "eos_token_id": 2,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 1024,
  "initializer_range": 0.02,
  "intermediate_size": 4096,
  "layer_norm_eps": 1e-05,
  "max_position_embeddings": 514,
  "model_type": "roberta",
  "nclass": 2,
  "new_tok_size": 32004,
  "num_attention_heads": 16,
  "num_hidden_layers": 24,
  "pad_token_id": 1,
  "position_embedding_type": "absolute",
  "tokenizer_class": "BertTokenizer",
  "transformers_version": "4.12.5",
  "type_vocab_size": 1,
  "use_cache": true,
  "vocab_size": 32000
}



In [12]:
def train_epoch(epoch, model, train_loader, optimizer, scheduler):
    model.train()
    total_loss = 0
    cor = 0
    n_sample = 0
    s = time.time()
    criterion = nn.CrossEntropyLoss()

    for data, target in tqdm(train_loader):
        item = {key: val.to(device) for key, val in data.items()}
        target = target.to(device)

        logits = model(**item)
        loss = criterion(logits, target)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        with torch.no_grad():
            preds = torch.argmax(logits, dim=-1)

        total_loss += loss.item()
        cor += (preds == target).sum().item()
        n_sample += len(target)

        print(f"{cor}/{n_sample}", end='\r')

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    print(
        f"[Epoch {epoch}] Train loss: {loss_avg:.3f}, acc: {acc*100:.2f}, time: {time.time()-s:.1f}s"
    )
    return acc


def validate(epoch, model, val_loader, verbose=True):
    model.eval()
    total_loss = 0
    cor = 0
    n_sample = 0
    criterion = nn.CrossEntropyLoss()
    pred_all = []
    
    with torch.no_grad():
        for data, target in val_loader:
            item = {key: val.to(device) for key, val in data.items()}
            target = target.to(device)

            logits = model(**item)
            loss = criterion(logits, target)
            preds = torch.argmax(logits, dim=-1)
            pred_all.append(preds)

            total_loss += loss.item()
            cor += (preds == target).sum().item()
            n_sample += len(target)

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    pred_all = torch.cat(pred_all)
    
    if verbose:
        print(f"[Epoch {epoch}] Valid loss: {loss_avg:.3f}, acc: {acc*100:.2f}")
    return acc, pred_all


def train(idx, num_epochs, lr, train_loader, val_loader, tokenizer):
    print(f"Start trining {idx}th model")
    model = Roberta(config, model_name).to(device)


    optimizer = AdamW(model.parameters(), lr=lr)
    scheduler = transformers.get_scheduler("linear",
                                           optimizer=optimizer,
                                           num_warmup_steps=num_epochs * len(train_loader) // 10,
                                           num_training_steps=num_epochs * len(train_loader))
    best_acc = 0
    for epoch in range(num_epochs):
        train_acc = train_epoch(epoch, model, train_loader, optimizer, scheduler)
        val_acc, _ = validate(epoch, model, val_loader)
        if val_acc > best_acc:
            best_acc = val_acc

            model_to_save = model.module if hasattr(model, "module") else model

            if not os.path.exists( os.path.join(save_dir, f'{idx}') ):
                os.makedirs( os.path.join(save_dir, f'{idx}') )

            model_to_save.save_pretrained(os.path.join(save_dir, f'{idx}'))
            
    print(f"Training finish! Best validation accuracy: {best_acc*100:.2f}\n")

### [WIC] 실행 결과 (validation: 93.31%)
 - 각 모델은 대략 91%~92.5% 정도의 결과를 보였습니다.
 - 6번 실행하여 이후에 앙상블을 사용합니다.

In [13]:
lr = 8e-6
num_epochs = 10
ensen_num=6

In [14]:
for i in range(ensen_num):
    train(i, num_epochs, lr, train_loader, val_loader, tokenizer)

Start trining 0th model


Downloading:   0%|          | 0.00/1.25G [00:00<?, ?B/s]

  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 0] Train loss: 0.127, acc: 73.70, time: 429.8s
[Epoch 0] Valid loss: 0.084, acc: 87.82


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 1] Train loss: 0.064, acc: 90.49, time: 429.7s
[Epoch 1] Valid loss: 0.071, acc: 88.16


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 2] Train loss: 0.026, acc: 96.32, time: 429.9s
[Epoch 2] Valid loss: 0.073, acc: 91.17


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 3] Train loss: 0.012, acc: 98.46, time: 430.1s
[Epoch 3] Valid loss: 0.093, acc: 90.14


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 4] Train loss: 0.007, acc: 99.01, time: 429.5s
[Epoch 4] Valid loss: 0.090, acc: 91.25


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 5] Train loss: 0.005, acc: 99.48, time: 432.8s
[Epoch 5] Valid loss: 0.106, acc: 90.91


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 6] Train loss: 0.002, acc: 99.81, time: 432.7s
[Epoch 6] Valid loss: 0.117, acc: 91.42


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 7] Train loss: 0.001, acc: 99.91, time: 432.9s
[Epoch 7] Valid loss: 0.133, acc: 92.02


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 8] Train loss: 0.000, acc: 99.99, time: 433.2s
[Epoch 8] Valid loss: 0.148, acc: 92.28


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 9] Train loss: 0.001, acc: 99.94, time: 432.8s
[Epoch 9] Valid loss: 0.142, acc: 92.28
Training finish! Best validation accuracy: 92.28

Start trining 1th model


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 0] Train loss: 0.127, acc: 72.66, time: 432.9s
[Epoch 0] Valid loss: 0.085, acc: 85.42


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 1] Train loss: 0.057, acc: 91.29, time: 433.7s
[Epoch 1] Valid loss: 0.076, acc: 88.59


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 2] Train loss: 0.021, acc: 97.11, time: 433.5s
[Epoch 2] Valid loss: 0.069, acc: 90.22


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 3] Train loss: 0.011, acc: 98.66, time: 432.9s
[Epoch 3] Valid loss: 0.103, acc: 88.85


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 4] Train loss: 0.006, acc: 99.37, time: 433.1s
[Epoch 4] Valid loss: 0.104, acc: 90.82


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 5] Train loss: 0.004, acc: 99.57, time: 433.3s
[Epoch 5] Valid loss: 0.089, acc: 90.91


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 6] Train loss: 0.002, acc: 99.81, time: 432.8s
[Epoch 6] Valid loss: 0.101, acc: 92.54


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 7] Train loss: 0.000, acc: 99.95, time: 433.1s
[Epoch 7] Valid loss: 0.130, acc: 91.60


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 8] Train loss: 0.001, acc: 99.94, time: 433.1s
[Epoch 8] Valid loss: 0.140, acc: 91.60


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 9] Train loss: 0.000, acc: 99.99, time: 433.3s
[Epoch 9] Valid loss: 0.130, acc: 92.20
Training finish! Best validation accuracy: 92.54

Start trining 2th model


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 0] Train loss: 0.130, acc: 72.73, time: 433.2s
[Epoch 0] Valid loss: 0.091, acc: 84.73


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 1] Train loss: 0.056, acc: 91.62, time: 433.4s
[Epoch 1] Valid loss: 0.071, acc: 88.34


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 2] Train loss: 0.022, acc: 97.02, time: 433.3s
[Epoch 2] Valid loss: 0.079, acc: 90.14


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 3] Train loss: 0.010, acc: 98.50, time: 433.1s
[Epoch 3] Valid loss: 0.096, acc: 90.31


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 4] Train loss: 0.006, acc: 99.16, time: 433.2s
[Epoch 4] Valid loss: 0.095, acc: 91.08


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 5] Train loss: 0.003, acc: 99.64, time: 434.0s
[Epoch 5] Valid loss: 0.121, acc: 90.48


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 6] Train loss: 0.002, acc: 99.74, time: 433.7s
[Epoch 6] Valid loss: 0.107, acc: 91.17


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 7] Train loss: 0.002, acc: 99.83, time: 433.5s
[Epoch 7] Valid loss: 0.114, acc: 91.85


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 8] Train loss: 0.000, acc: 100.00, time: 433.7s
[Epoch 8] Valid loss: 0.128, acc: 92.11


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 9] Train loss: 0.000, acc: 99.96, time: 433.2s
[Epoch 9] Valid loss: 0.134, acc: 91.94
Training finish! Best validation accuracy: 92.11

Start trining 3th model


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 0] Train loss: 0.126, acc: 73.52, time: 434.0s
[Epoch 0] Valid loss: 0.076, acc: 87.99


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 1] Train loss: 0.059, acc: 90.95, time: 433.2s
[Epoch 1] Valid loss: 0.070, acc: 89.62


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 2] Train loss: 0.023, acc: 96.77, time: 433.9s
[Epoch 2] Valid loss: 0.076, acc: 90.22


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 3] Train loss: 0.013, acc: 98.26, time: 433.5s
[Epoch 3] Valid loss: 0.132, acc: 88.25


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 4] Train loss: 0.007, acc: 99.04, time: 433.3s
[Epoch 4] Valid loss: 0.072, acc: 91.42


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 5] Train loss: 0.003, acc: 99.65, time: 434.6s
[Epoch 5] Valid loss: 0.127, acc: 90.82


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 6] Train loss: 0.002, acc: 99.73, time: 433.4s
[Epoch 6] Valid loss: 0.166, acc: 87.82


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 7] Train loss: 0.001, acc: 99.86, time: 433.6s
[Epoch 7] Valid loss: 0.122, acc: 90.74


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 8] Train loss: 0.001, acc: 99.92, time: 433.9s
[Epoch 8] Valid loss: 0.118, acc: 92.20


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 9] Train loss: 0.000, acc: 99.94, time: 433.7s
[Epoch 9] Valid loss: 0.122, acc: 92.11
Training finish! Best validation accuracy: 92.20

Start trining 4th model


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 0] Train loss: 0.126, acc: 74.42, time: 432.7s
[Epoch 0] Valid loss: 0.085, acc: 84.56


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 1] Train loss: 0.058, acc: 91.34, time: 433.5s
[Epoch 1] Valid loss: 0.064, acc: 89.88


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 2] Train loss: 0.022, acc: 96.97, time: 432.9s
[Epoch 2] Valid loss: 0.062, acc: 90.74


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 3] Train loss: 0.011, acc: 98.70, time: 433.6s
[Epoch 3] Valid loss: 0.079, acc: 91.51


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 4] Train loss: 0.007, acc: 99.06, time: 433.2s
[Epoch 4] Valid loss: 0.077, acc: 91.25


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 5] Train loss: 0.003, acc: 99.54, time: 433.6s
[Epoch 5] Valid loss: 0.106, acc: 91.08


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 6] Train loss: 0.002, acc: 99.73, time: 433.4s
[Epoch 6] Valid loss: 0.113, acc: 91.60


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 7] Train loss: 0.001, acc: 99.88, time: 432.9s
[Epoch 7] Valid loss: 0.108, acc: 92.02


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 8] Train loss: 0.000, acc: 99.94, time: 433.9s
[Epoch 8] Valid loss: 0.119, acc: 92.28


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 9] Train loss: 0.000, acc: 99.97, time: 433.4s
[Epoch 9] Valid loss: 0.125, acc: 92.28
Training finish! Best validation accuracy: 92.28

Start trining 5th model


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 0] Train loss: 0.129, acc: 72.23, time: 433.1s
[Epoch 0] Valid loss: 0.111, acc: 80.02


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 1] Train loss: 0.058, acc: 91.21, time: 433.8s
[Epoch 1] Valid loss: 0.059, acc: 90.14


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 2] Train loss: 0.022, acc: 97.11, time: 433.7s
[Epoch 2] Valid loss: 0.081, acc: 90.48


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 3] Train loss: 0.012, acc: 98.40, time: 433.1s
[Epoch 3] Valid loss: 0.108, acc: 89.79


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 4] Train loss: 0.007, acc: 99.23, time: 434.3s
[Epoch 4] Valid loss: 0.093, acc: 90.48


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 5] Train loss: 0.003, acc: 99.56, time: 433.6s
[Epoch 5] Valid loss: 0.134, acc: 90.22


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 6] Train loss: 0.002, acc: 99.73, time: 433.0s
[Epoch 6] Valid loss: 0.129, acc: 90.99


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 7] Train loss: 0.001, acc: 99.83, time: 434.0s
[Epoch 7] Valid loss: 0.120, acc: 91.60


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 8] Train loss: 0.001, acc: 99.94, time: 433.5s
[Epoch 8] Valid loss: 0.121, acc: 91.51


  0%|          | 0/1937 [00:00<?, ?it/s]

[Epoch 9] Train loss: 0.000, acc: 99.99, time: 433.1s
[Epoch 9] Valid loss: 0.126, acc: 91.68
Training finish! Best validation accuracy: 91.68



In [20]:
def validate_ensemble(val_loader, answer, idx_max=10):
    pred_ensemble = []
    for idx in range(idx_max):
        model = Roberta.from_pretrained(os.path.join(save_dir, f'{idx}'), model_name)
        model.to(device)
        acc, pred_all = validate('best', model, val_loader, verbose=False)
        print(f"Load {idx}th model (acc: {acc*100:.2f})")
        if acc >= 0.87:
            pred_ensemble.append(pred_all)
        
    pred_ensemble = torch.stack(pred_ensemble, dim=-1).float()
    pred_ensemble = (pred_ensemble.mean(-1) >= 0.5).long().to(answer.device)
    acc_ensemble = (pred_ensemble == answer).sum() / len(answer)
    print(f"\nEnsemble accuracy: {acc_ensemble*100:.2f}")

In [21]:
answer = torch.tensor(val_dataset.labels)

validate_ensemble(val_loader, answer, idx_max=ensen_num)

Load 0th model (acc: 92.28)
Load 1th model (acc: 92.54)
Load 2th model (acc: 92.11)
Load 3th model (acc: 92.20)
Load 4th model (acc: 92.28)
Load 5th model (acc: 91.68)

Ensemble accuracy: 93.31


## 3. 인과관계추론 (COPA)
아래 하이퍼파라미터 및 모델 세팅을 탐색하였음.
가장 성능이 좋은 세팅은 강조되어 있음. 

*전처리*
- Question이 원인인지, 결과인지에 따라 문장을 각각의 리스트로 재구성하여 전처리. 이에 따라 2개의 데이터셋으로 나누어 토크나이징 진행
- 나누어 진행된 tokeized sentence는 합쳐서 return
- 전처리의 경우, 원인 결과를 명시한 경우와 아닌 경우 각각을 실험한 결과 명시하지 않은 경우 성능이 더 좋았음(92.8 < **93.2**)

*모델*
- Model : **KLUE-RoBERTa(90~93%)** > Ko-ELECTRa-v3(88.4%)


*세팅*
- Max length & Batch combination : 100/8(92.2%), 100/16(92.8%), **150/5(93.2%)**, 150/8(92.4%), 200/5(92.8%), 200/8(92.8%)
- Epoch : 10
- Warm up : 10% training step
- Learning rate : 8e-6
- Optimizer : **AdamW(93.2%)** vs RAdam(93%)
- Scheduler : Linear Scheduler
* Experiment : https://github.com/codestella/nlp_final_project/tree/main/COPA_experiment



In [ ]:
import os
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
import transformers
import pandas as pd
from transformers import AutoTokenizer
from transformers import RobertaModel, RobertaConfig
from transformers import AdamW
import time
import argparse

transformers.logging.set_verbosity(40) # Turn off warning
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

save_dir = './result_copa'
if not os.path.exists(save_dir):
    os.makedirs(save_dir)

In [2]:
def load_data(dataset_dir):
    dataset = pd.read_csv(dataset_dir, delimiter='\t', names=['ID', 'sentence', 'question', '1', '2','answer'], header=0)
    dataset["label"] = dataset["answer"].astype(int) - 1

    new_sentence1_1 = []
    new_sentence1_2 = []
    new_sentence2_1 = []
    new_sentence2_2 = []

    for i in range(len(dataset)):
        s = dataset.iloc[i]['sentence']
        q = dataset.iloc[i]['question']
        s1 = dataset.iloc[i]['1']
        s2 = dataset.iloc[i]['2']
        lb = dataset.iloc[i]['label']

        if q == "결과":
            new_sentence1_1.append(s)
            new_sentence1_2.append(s1)  # 후보1
            new_sentence2_1.append(s)
            new_sentence2_2.append(s2)  # 후보2

        else:
            new_sentence1_1.append(s1) # 후보1 먼저
            new_sentence1_2.append(s)
            new_sentence2_1.append(s2) # 후보2 먼저 
            new_sentence2_2.append(s) 

    dataset["new_sentence1_1"] = new_sentence1_1
    dataset["new_sentence1_2"] = new_sentence1_2
    dataset["new_sentence2_1"] = new_sentence2_1
    dataset["new_sentence2_2"] = new_sentence2_2

    return dataset


class TensorDataset(Dataset):
    ''' Define Torch Dataset '''
    def __init__(self, tokenized_dataset, labels):
        self.tokenized_dataset = tokenized_dataset
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: val[idx] for key, val in self.tokenized_dataset.items()}
        label = self.labels[idx]
        return item, label

    def __len__(self):
        return len(self.labels)

In [ ]:
def tokenized_dataset(dataset, tokenizer) :
    sentence1_1 = dataset['new_sentence1_1'].tolist()
    sentence1_2 = dataset['new_sentence1_2'].tolist()
    sentence2_1 = dataset["new_sentence2_1"].tolist()
    sentence2_2 = dataset["new_sentence2_2"].tolist()

    # 결과인 경우
    tokenized_sentences = tokenizer(
      sentence1_1,
      sentence1_2,
      return_tensors  = 'pt',
      padding = True,
      truncation = True, 
      max_length = 150, 
      add_special_tokens = True,
      return_token_type_ids = True
    )

    # 원인인 경우
    tokenized_sentences2 = tokenizer(
      sentence2_1,
      sentence2_2,
      return_tensors  = 'pt',
      padding = True,
      truncation = True, 
      max_length = 150,  
      add_special_tokens = True,
      return_token_type_ids = True
    )

    # Sentence 하나로 합쳐 return 하기
    for key, value in tokenized_sentences2.items():
        tokenized_sentences[key+'2'] = value

    return tokenized_sentences

In [ ]:
model_type = "Roberta"
size = 'large'
model_name = f"klue/roberta-{size}"
tokenizer = AutoTokenizer.from_pretrained(model_name)

Downloading:   0%|          | 0.00/375 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/243k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/734k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/173 [00:00<?, ?B/s]

In [ ]:
class Roberta(RobertaModel):
    def __init__(self, config, model_name):
        super(Roberta, self).__init__(config)
        self.roberta = RobertaModel.from_pretrained(model_name, config=config)
        self.hdim = config.hidden_size
        self.nclass = config.nclass
        self.classifier = nn.Linear(self.hdim, self.nclass-1)

    def forward(self, input_ids, input_ids2, attention_mask, attention_mask2, **kwargs):
        outputs = self.roberta(input_ids, attention_mask=attention_mask)
        outputs2 = self.roberta(input_ids2, attention_mask=attention_mask2)
        h = outputs[0][:, 0, :]
        h2 = outputs2[0][:, 0, :]
        logits1 = self.classifier(h)
        logits2 = self.classifier(h2)
        logits = torch.cat([logits1, logits2], dim=1)
        return logits

config = RobertaConfig.from_pretrained(model_name)
config.nclass = 2

Downloading:   0%|          | 0.00/547 [00:00<?, ?B/s]

In [ ]:
# Define loader
if size == 'base':
    batch_size = 16
else:
    batch_size = 5
    
# Data path
base_path = './data'
train_dataset = load_data(os.path.join(base_path, 'SKT_COPA_Train.tsv'))
val_dataset = load_data(os.path.join(base_path, 'SKT_COPA_Dev.tsv'))

train_tokenized = tokenized_dataset(train_dataset, tokenizer)
val_tokenized = tokenized_dataset(val_dataset, tokenizer)

train_dataset = TensorDataset(train_tokenized, train_dataset['label'])
val_dataset = TensorDataset(val_tokenized, val_dataset['label'])
    
train_loader = DataLoader(train_dataset, batch_size=batch_size,drop_last=True, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=batch_size, drop_last=False, shuffle=False)

### [COPA] 전처리 예시
- input_ids에 후보 1 문장, input_ids2에 후보 2문장이 들어가 있는 형태로 구성되어 있음

In [ ]:
tokenizer.decode(train_tokenized['input_ids'][100])

'[CLS] 바닥에 기어다니는 바퀴벌레를 발견했다. [SEP] 모기향을 피웠다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]'

In [ ]:
tokenizer.decode(train_tokenized['input_ids2'][100])

'[CLS] 날아다니는 모기를 발견했다. [SEP] 모기향을 피웠다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]'

In [ ]:
tokenizer.decode(val_tokenized['input_ids'][0])

'[CLS] 전쟁이 시작되었다. [SEP] 병사들이 집으로 돌아왔다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]'

In [ ]:
tokenizer.decode(val_tokenized['input_ids2'][0])

'[CLS] 전쟁이 시작되었다. [SEP] 병사들이 전투에 파견되었다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]'

### [COPA] 학습 결과 (Validation: 93.2%)
* 최적 세팅에서 10번 반복한 결과 90.8% ~ 92.8% 의 결과를 얻음
* 각 실험은 약 23분 정도 소요 (2.3 min/epoch)
* 본 모델을 앙상블하여 최종 93.2%의 validation accuracy 달성.

In [ ]:
def train_epoch(epoch, model, train_loader, optimizer, scheduler):
    model.train()
    total_loss = 0
    cor = 0
    n_sample = 0
    s = time.time()
    criterion = nn.CrossEntropyLoss()

    for data, target in train_loader:
        item = {key: val.to(device) for key, val in data.items()}
        target = target.to(device)

        logits = model(**item)
        loss = criterion(logits, target)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        with torch.no_grad():
            preds = torch.argmax(logits, dim=-1)

        total_loss += loss.item()
        cor += (preds == target).sum().item()
        n_sample += len(target)

        print(f"{cor}/{n_sample}", end='\r')

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    print(
        f"[Epoch {epoch}] Train loss: {loss_avg:.3f}, acc: {acc*100:.2f}, time: {time.time()-s:.1f}s"
    )
    return acc


def validate(epoch, model, val_loader, verbose=True):
    model.eval()
    total_loss = 0
    cor = 0
    n_sample = 0
    criterion = nn.CrossEntropyLoss()
    pred_all = []
    
    with torch.no_grad():
        for data, target in val_loader:
            item = {key: val.to(device) for key, val in data.items()}
            target = target.to(device)

            logits = model(**item)
            loss = criterion(logits, target)
            preds = torch.argmax(logits, dim=-1)
            pred_all.append(preds)

            total_loss += loss.item()
            cor += (preds == target).sum().item()
            n_sample += len(target)

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    pred_all = torch.cat(pred_all)
    
    if verbose:
        print(f"[Epoch {epoch}] Valid loss: {loss_avg:.3f}, acc: {acc*100:.2f}")
    return acc, pred_all


def train(idx, num_epochs, lr, train_loader, val_loader):
    print(f"Start trining {idx}th model")
    model = Roberta(config, model_name).to(device)
    optimizer = AdamW(model.parameters(), lr=lr)
    scheduler = transformers.get_scheduler("linear",
                                           optimizer=optimizer,
                                           num_warmup_steps=num_epochs * len(train_loader) // 10,
                                           num_training_steps=num_epochs * len(train_loader))
    best_acc = 0
    for epoch in range(num_epochs):
        train_acc = train_epoch(epoch, model, train_loader, optimizer, scheduler)
        val_acc, _ = validate(epoch, model, val_loader)
        if val_acc > best_acc:
            best_acc = val_acc

            model_to_save = model.module if hasattr(model, "module") else model
            model_to_save.save_pretrained(os.path.join(save_dir, f'{idx}'))
            
    print(f"Training finish! Best validation accuracy: {best_acc*100:.2f}\n")

In [ ]:
lr = 8e-6
num_epochs = 10

In [ ]:
for i in range(10):
    train(i, num_epochs, lr, train_loader, val_loader)

Start trining 0th model


Downloading:   0%|          | 0.00/1.25G [00:00<?, ?B/s]

[Epoch 0] Train loss: 0.107, acc: 70.36, time: 140.0s
[Epoch 0] Valid loss: 0.061, acc: 86.60
[Epoch 1] Train loss: 0.048, acc: 90.10, time: 139.6s
[Epoch 1] Valid loss: 0.056, acc: 89.20
[Epoch 2] Train loss: 0.018, acc: 96.62, time: 139.7s
[Epoch 2] Valid loss: 0.065, acc: 91.60
[Epoch 3] Train loss: 0.008, acc: 98.64, time: 139.6s
[Epoch 3] Valid loss: 0.075, acc: 91.00
[Epoch 4] Train loss: 0.005, acc: 99.09, time: 139.7s
[Epoch 4] Valid loss: 0.062, acc: 92.20
[Epoch 5] Train loss: 0.003, acc: 99.81, time: 139.7s
[Epoch 5] Valid loss: 0.071, acc: 91.00
[Epoch 6] Train loss: 0.001, acc: 99.87, time: 139.7s
[Epoch 6] Valid loss: 0.073, acc: 91.20
[Epoch 7] Train loss: 0.001, acc: 99.84, time: 139.7s
[Epoch 7] Valid loss: 0.073, acc: 91.80
[Epoch 8] Train loss: 0.001, acc: 100.00, time: 139.7s
[Epoch 8] Valid loss: 0.077, acc: 91.00
[Epoch 9] Train loss: 0.001, acc: 99.94, time: 139.7s
[Epoch 9] Valid loss: 0.077, acc: 91.60
Training finish! Best validation accuracy: 92.20

Start tri

In [ ]:
def validate_ensemble(val_loader, answer, idx_max=10):
    pred_ensemble = []
    for idx in range(idx_max):
        model = Roberta.from_pretrained(os.path.join(save_dir, f'{idx}'), model_name)
        model.to(device)
        acc, pred_all = validate('best', model, val_loader, verbose=False)
        print(f"Load {idx}th model (acc: {acc*100:.2f})")
        if acc >= 0.85:
            pred_ensemble.append(pred_all)
        
    pred_ensemble = torch.stack(pred_ensemble, dim=-1).float()
    pred_ensemble = (pred_ensemble.mean(-1) >= 0.5).long().to(answer.device)
    acc_ensemble = (pred_ensemble == answer).sum() / len(answer)
    print(f"\nEnsemble accuracy: {acc_ensemble*100:.2f}")

In [ ]:
answer = torch.tensor(val_dataset.labels)
validate_ensemble(val_loader, answer, idx_max=10)

Load 0th model (acc: 92.20)
Load 1th model (acc: 91.80)
Load 2th model (acc: 91.00)
Load 3th model (acc: 92.80)
Load 4th model (acc: 91.60)
Load 5th model (acc: 90.80)
Load 6th model (acc: 92.40)
Load 7th model (acc: 91.00)
Load 8th model (acc: 91.40)
Load 9th model (acc: 90.80)

Ensemble accuracy: 93.20


## 4. 문법성 판단 (CoLA)

In [25]:
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import transformers
import pandas as pd
from transformers import AutoTokenizer
from transformers import ElectraConfig, ElectraModel
from transformers import AdamW
import time
import argparse
transformers.logging.set_verbosity(40) # Turn off warning

save_dir = './result_gram'
if not os.path.exists(save_dir):
    os.makedirs(save_dir)

### [CoLA] Load data
- Train, Dev 데이터가 base_path에 들어있어야 합니다. (default: './data')

In [27]:
def load_data(path, tokenizer):
    dataset = pd.read_csv(path, delimiter='\t', names=['source', 'acceptability_label', 'source_annotation', 'sentence'], header=0)
    dataset["label"] = dataset["acceptability_label"].astype(int)
    
    sentence = dataset['sentence'].tolist()
    
    tokenized = tokenizer(sentence,
    return_tensors="pt",
    padding=True,
    truncation=True,
    max_length=150,
    add_special_tokens=True,
    return_token_type_ids = True)
    
    return dataset, tokenized


class TensorDataset(Dataset):
    def __init__(self, tokenized_dataset, labels):
        self.tokenized_dataset = tokenized_dataset
        self.labels = labels

    def __getitem__(self, idx):
        item = {key: val[idx] for key, val in self.tokenized_dataset.items()}
        label = self.labels[idx]
        return item, label

    def __len__(self):
        return len(self.labels)

In [28]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

model_type = "Electra"
model_name = f"monologg/koelectra-base-v3-discriminator"
tokenizer = AutoTokenizer.from_pretrained(model_name)

In [29]:
base_path = './data'
train_dataset, train_tokenized = load_data(os.path.join(base_path, 'NIKL_CoLA_train.tsv'), tokenizer)
val_dataset, val_tokenized = load_data(os.path.join(base_path, 'NIKL_CoLA_dev.tsv'), tokenizer)

train_dataset = TensorDataset(train_tokenized, train_dataset['label'])
val_dataset = TensorDataset(val_tokenized, val_dataset['label'])

batch_size = 32

train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)

In [30]:
# 데이터 예시
tokenizer.decode(train_tokenized['input_ids'][0])

'[CLS] 높은 달이 떴다. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]'

### [CoLA] Load pretrained model
- 모델은 monologg/koelectra-base-v3-discriminator를 사용하였습니다. (https://github.com/monologg/KoELECTRA) 

**학습 방법**:  

다음의  logit 값을 더하여 최종 분류를 하였습니다. 
- cls token에 해당하는 hidden feature에 3-layer classifier 
- average hidden feature에 classifier  

In [31]:
class Electra(ElectraModel):
    def __init__(self, config, model_name):
        super(Electra, self).__init__(config)
        self.electra = ElectraModel.from_pretrained(model_name, config=config)
        self.hdim = config.hidden_size
        self.nclass = config.nclass
        self.drop = nn.Dropout(p=0.2)
        self.layer1 = nn.Linear(self.hdim, 256) #258
        self.layer2 = nn.Linear(256, 128)
        self.layer3 = nn.Linear(128, self.nclass)
        self.layer4 = nn.Linear(self.hdim, self.nclass)

    def forward(self, input_ids, attention_mask, **kwargs):
        outputs = self.electra(input_ids, attention_mask=attention_mask)
        h = outputs[0][:, 0, :] 
        h = self.drop(h)
        h = F.relu(self.layer1(h))
        h = F.relu(self.layer2(h))
        logits = self.layer3(h) 
        result = logits + self.layer4(torch.mean(outputs[0], 1)) 
        return result

config = ElectraConfig.from_pretrained(model_name)
config.nclass = 2

### [CoLA] Finetune model (Validation: 74.85% (acc), 55.65% (MCC))
- Learning rate: 1e-6
- Epoch: 80
- 학습 과정 중 Best 모델을 저장하여 평가하였습니다. 

In [32]:
def train_epoch(epoch, model, train_loader, optimizer, scheduler):
    model.train()
    total_loss = 0
    cor = 0
    n_sample = 0
    TP = 0
    FP = 0
    TN = 0
    FN = 0
    eps = 1e-9
    s = time.time()
    criterion = nn.CrossEntropyLoss()
    add_MCC =0
    
    for data, target in train_loader:
        item = {key: val.to(device) for key, val in data.items()}
        target = target.to(device)

        logits = model(**item)
        loss = criterion(logits, target)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        with torch.no_grad():
            preds = torch.argmax(logits, dim=-1)

        total_loss += loss.item()
        cor += (preds == target).sum().item()
        n_sample += len(target)

        print(f"{cor}/{n_sample}", end='\r')
        
        TRUE = (target == preds)
        FALSE = (target != preds)

        TP += (TRUE * preds).sum().item()
        TN += (TRUE * (preds==0)).sum().item()
        FP += (FALSE * preds).sum().item()
        FN += (FALSE * (preds==0)).sum().item()

        MCC = ((TP*TN) - (FP*FN)) / (((TP+FP+eps)*(TP+FN+eps)*(TN+FP+eps)*(TN+FN+eps))**0.5)
        add_MCC += MCC

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    end_MCC = add_MCC / len(train_loader)
    print(
        f"[Epoch {epoch}] Train loss: {loss_avg:.3f}, mcc: {end_MCC*100:.2f}, acc: {acc*100:.2f}, time: {time.time()-s:.1f}s"
    ) 
    return end_MCC


def validate(epoch, model, val_loader, verbose=True):
    model.eval()
    total_loss = 0
    cor = 0
    n_sample = 0
    TP = 0
    FP = 0
    TN = 0
    FN = 0
    eps = 1e-9
    criterion = nn.CrossEntropyLoss()
    pred_all = []
    add_MCC =0
    
    with torch.no_grad():
        for data, target in val_loader:
            item = {key: val.to(device) for key, val in data.items()}
            target = target.to(device)

            logits = model(**item)
            loss = criterion(logits, target)
            preds = torch.argmax(logits, dim=-1)
            pred_all.append(preds)

            total_loss += loss.item()
            cor += (preds == target).sum().item() 
            n_sample += len(target)
            
            TRUE = (target == preds)
            FALSE = (target != preds)

            TP += (TRUE * preds).sum().item()
            TN += (TRUE * (preds==0)).sum().item()
            FP += (FALSE * preds).sum().item()
            FN += (FALSE * (preds==0)).sum().item()

            MCC = ((TP*TN) - (FP*FN)) / (((TP+FP+eps)*(TP+FN+eps)*(TN+FP+eps)*(TN+FN+eps))**0.5)
            add_MCC += MCC

    loss_avg = total_loss / n_sample
    acc = cor / n_sample
    pred_all = torch.cat(pred_all)
    end_MCC = add_MCC / len(val_loader)

    if verbose:
        print(f"[Epoch {epoch}] Valid loss: {loss_avg:.3f}, mcc: {end_MCC*100:.2f}, acc: {acc*100:.2f}") 
    return end_MCC, pred_all


def train(idx, num_epochs, lr, train_loader, val_loader):
    print(f"Start trining {idx}th model")
    model = Electra(config, model_name).to(device)
    optimizer = AdamW(model.parameters(), lr=lr)
    scheduler = transformers.get_scheduler("linear",
                                           optimizer=optimizer,
                                           num_warmup_steps=num_epochs * len(train_loader) // 10,
                                           num_training_steps=num_epochs * len(train_loader))
    best_mcc = 0
    for epoch in range(num_epochs):
        train_mcc = train_epoch(epoch, model, train_loader, optimizer, scheduler)
        val_mcc, _ = validate(epoch, model, val_loader)
        if val_mcc > best_mcc:
            best_mcc = val_mcc

            model_to_save = model.module if hasattr(model, "module") else model
            model_to_save.save_pretrained(os.path.join(save_dir, f'{idx}'))
            
    print(f"Training finish! Best validation mcc: {best_mcc*100:.2f}\n")

In [33]:
lr = 1e-6 
num_epochs = 80

In [34]:
train(0, num_epochs, lr, train_loader, val_loader)
# We run 10 times in the results below

Start trining 0th model
[Epoch 0] Train loss: 0.022, mcc: -0.91, acc: 49.71, time: 44.4s
[Epoch 0] Valid loss: 0.022, mcc: -2.24, acc: 52.46
[Epoch 1] Train loss: 0.022, mcc: -1.91, acc: 50.13, time: 44.4s
[Epoch 1] Valid loss: 0.022, mcc: 23.38, acc: 60.88
[Epoch 2] Train loss: 0.021, mcc: 14.94, acc: 60.47, time: 43.7s
[Epoch 2] Valid loss: 0.021, mcc: 41.50, acc: 68.60
[Epoch 3] Train loss: 0.020, mcc: 35.36, acc: 68.46, time: 43.7s
[Epoch 3] Valid loss: 0.019, mcc: 42.64, acc: 69.64
[Epoch 4] Train loss: 0.018, mcc: 42.63, acc: 71.81, time: 44.5s
[Epoch 4] Valid loss: 0.017, mcc: 46.93, acc: 72.24
[Epoch 5] Train loss: 0.017, mcc: 47.28, acc: 73.97, time: 43.4s
[Epoch 5] Valid loss: 0.017, mcc: 47.14, acc: 72.79
[Epoch 6] Train loss: 0.016, mcc: 52.36, acc: 76.05, time: 44.1s
[Epoch 6] Valid loss: 0.016, mcc: 49.23, acc: 74.06
[Epoch 7] Train loss: 0.015, mcc: 53.28, acc: 77.02, time: 45.6s
[Epoch 7] Valid loss: 0.016, mcc: 51.54, acc: 75.00
[Epoch 8] Train loss: 0.015, mcc: 56.32,

In [46]:
def validate_last(val_loader, answer):
    model = Electra.from_pretrained(os.path.join(save_dir, f'0'), model_name)
    model.to(device)
    mcc, pred_all = validate('best', model, val_loader, verbose=False)
    print(f"Load final model (mcc: {mcc*100:.2f})")

In [47]:
answer = torch.tensor(val_dataset.labels)

validate_last(val_loader, answer)

Load final model (mcc: 55.65)
